scripts/stakingSimulator.js

```diff
@@ -0,0 +1,828 @@
+            isValid &= (assertSameStakerDataInvariants(initState.newRep, newState.newRep, actionDoneInSameEpoch));
+    }
+    return isValid;
+}
+
+function assertSameStakerDataInvariants(initStakerData, newStakerData, actionDoneInSameEpoch) {
+    let isValid = true;
+
+    if (actionDoneInSameEpoch) {
+        isValid &= assertSameStakerData(initStakerData.dataCurEpoch, newStakerData.dataCurEpoch);
+        if (!isValid) {
+            console.log(`dataCurEpochs don't match`);
+        }
+        isValid &= assertSameStakerData(newStakerData.dataNextEpoch, newStakerData.latestData);
+        if (!isValid) {
+            console.log(`newStaker latestData & dataNextEpoch don't match`);
+        }
+    } else {
+        isValid &= assertSameStakerData(initStakerData.dataNextEpoch, newStakerData.dataCurEpoch);
+        if (!isValid) {
+            console.log(`Diff epochs: newStaker dataCurEpoch & initStakerData.dataNextEpoch don't match`);
+        }
+        isValid &= assertSameStakerData(newStakerData.dataNextEpoch, newStakerData.latestData);
+        if (!isValid) {
+            console.log(`Diff epochs: newStaker dataNextEpoch & latestData don't match`);
+        }
+    }
+    return isValid;
+}
+
+function assertSameStakerData(stakerData1, stakerData2) {
+    let isValid = true;
+    isValid &= (stakerData1.stake.eq(stakerData2.stake));
+    if (!isValid) {
+        console.log(`stakes don't match`);
+        console.log(`${stakerData1.stake.toString()} != ${stakerData2.stake.toString()}`);
+    }
+    isValid &= (stakerData1.dStake.eq(stakerData2.dStake));
+    if (!isValid) {
+        console.log(`delegated stakes don't match`);
+        console.log(`${stakerData1.dStake.toString()} != ${stakerData2.dStake.toString()}`);
+    }
+    isValid &= (stakerData1.dAddress == stakerData2.dAddress);
+    if (!isValid) {
+        console.log(`delegated addresses don't match`);
+        console.log(`${stakerData1.dAddress.toString()} != ${stakerData2.dAddress.toString()}`);
+    }
+    return isValid;
+}
+
+function assertSameDataStruct(initStaker, newStaker) {
+    let isValid = true;
+    isValid &= assertSameStakerData(initStaker.dataCurEpoch, newStaker.dataCurEpoch);
+    isValid &= assertSameStakerData(initStaker.dataNextEpoch, newStaker.dataNextEpoch);
+    isValid &= assertSameStakerData(initStaker.latestData, newStaker.latestData);
+    return isValid;
+}
+
+function assertSameDelegateAddresses(initState, newState) {
+    let isValid = true;
+    let sameEpoch = initState.epochNum.eq(newState.epochNum);
+    isValid &= assertSameDelegateAddress(initState.staker, newState.staker, sameEpoch);
+    isValid &= assertSameDelegateAddress(initState.oldRep, newState.oldRep, sameEpoch);
+    isValid &= assertSameDelegateAddress(initState.newRep, newState.newRep, sameEpoch);
+    return isValid;
+}
+
+function assertSameDelegateAddress(initStaker, newStaker, sameEpoch) {
+    let isValid = true;
```

Comment:
like we said.
if false. probably can print and return.

Reply:
Yup, it's like this now.